src/backend/opencl/memory.cpp

```diff
@@ -56,42 +57,69 @@ template<typename T>
 unique_ptr<cl::Buffer, function<void(cl::Buffer *)>> memAlloc(
     const size_t &elements) {
     // TODO: make memAlloc aware of array shapes
-    dim4 dims(elements);
-    void *ptr = memoryManager().alloc(false, 1, dims.get(), sizeof(T));
-    auto *buf = static_cast<cl::Buffer *>(ptr);
-    return unique_ptr<cl::Buffer, function<void(cl::Buffer *)>>(buf,
-                                                                bufferFree);
+    if (elements) {
+        dim4 dims(elements);
+        void *ptr = memoryManager().alloc(false, 1, dims.get(), sizeof(T));
+        auto buf  = static_cast<cl_mem>(ptr);
+        cl::Buffer *bptr = new cl::Buffer(buf, true);
+        return unique_ptr<cl::Buffer, function<void(cl::Buffer *)>>(bptr,
+                                                                    bufferFree);
+    } else {
+        return unique_ptr<cl::Buffer, function<void(cl::Buffer *)>>(nullptr,
+                                                                    bufferFree);
+    }
 }
 
 void *memAllocUser(const size_t &bytes) {
     dim4 dims(bytes);
     void *ptr = memoryManager().alloc(true, 1, dims.get(), 1);
-    return ptr;
+    auto buf  = static_cast<cl_mem>(ptr);
+    return new cl::Buffer(buf, true);
 }
 
 template<typename T>
 void memFree(T *ptr) {
-    return memoryManager().unlock(static_cast<void *>(ptr), false);
+    cl::Buffer *buf = reinterpret_cast<cl::Buffer *>(ptr);
+    cl_mem mem      = static_cast<cl_mem>((*buf)());
+    delete buf;
+    return memoryManager().unlock(static_cast<void *>(mem), false);
 }
 
-void memFreeUser(void *ptr) { memoryManager().unlock(ptr, true); }
+void memFreeUser(void *ptr) {
+    cl::Buffer *buf = static_cast<cl::Buffer *>(ptr);
+    cl_mem mem      = (*buf)();
+    delete buf;
+    memoryManager().unlock(mem, true);
+}
 
 cl::Buffer *bufferAlloc(const size_t &bytes) {
     dim4 dims(bytes);
-    void *ptr = memoryManager().alloc(false, 1, dims.get(), 1);
-    return static_cast<cl::Buffer *>(ptr);
+    if (bytes) {
+        void *ptr       = memoryManager().alloc(false, 1, dims.get(), 1);
+        cl_mem mem      = static_cast<cl_mem>(ptr);
+        cl::Buffer *buf = new cl::Buffer(mem, true);
+        return buf;
+    } else {
+        return nullptr;
+    }
 }
 
 void bufferFree(cl::Buffer *buf) {
-    return memoryManager().unlock(static_cast<void *>(buf), false);
+    if (buf) {
+        cl_mem mem = (*buf)();
+        delete buf;
+        memoryManager().unlock(static_cast<void *>(mem), false);
+    }
 }
 
-void memLock(const void *ptr) {
-    memoryManager().userLock(const_cast<void *>(ptr));
+void memLock(const cl::Buffer *ptr) {
```

Comment:
I think cl::Buffers are fine internally as long as the native functions are using cl_mem. The mem* functions are used internally and they should be returning cl::Buffers because that is most convenient for our code base. The reason I moved to the nativeAlloc interface is because the memory manager should be compatible with the base OpenCL specification. The C++ interface is not stable enough for that.